File: bindings/ts/src/embedded.ts
// kevy-embedded: the in-process Store the mem:// / file:// URLs use
// (contract §5.2). One EmbeddedDb wraps a low-level FFI handle and parses
// raw RESP bytes into typed Reply values. This is the conformance-minimum
// embedded surface — cmd(argv), scalar get/set, and polled/blocking pub/sub
// — and is exported publicly for callers who want the store directly.

import { decodeReply, toBytes, type Data, type Reply } from "./resp.ts";
import { loadEngine, type RawHandle, type RawSub } from "./ffi.ts";
import { ClosedError } from "./errors.ts";
import { registryKey, type Target } from "./url.ts";

/** An embedded subscription handle: poll with next(), block with wait(). */
export class EmbSub {
  #raw: RawSub | null;
  constructor(raw: RawSub) {
    this.#raw = raw;
  }
  /** Drain one queued frame without blocking (null if none). */
  next(): Reply | null {
    if (!this.#raw) throw new ClosedError("subscription closed");
    const bytes = this.#raw.nextRaw();
    return bytes ? decodeReply(bytes) : null;
  }
  /** Block up to timeoutMs (0 = forever) for one frame (null on timeout). */
  wait(timeoutMs: number): Reply | null {
    if (!this.#raw) throw new ClosedError("subscription closed");
    const bytes = this.#raw.waitRaw(timeoutMs);
    return bytes ? decodeReply(bytes) : null;
  }
  close(): void {
    this.#raw?.close();
    this.#raw = null;
  }
}

/** An open in-process store (contract §5.2). Close exactly once. */
export class EmbeddedDb {
  #raw: RawHandle | null;
  private constructor(raw: RawHandle) {
    this.#raw = raw;
  }

  /** Open a persistent store rooted at dir, replaying its log on open. */
  static open(dir: string): EmbeddedDb {
    return new EmbeddedDb(loadEngine().open(dir));
  }
  /** Open a pure in-memory store — nothing survives the process. */
  static openMem(): EmbeddedDb {
    return new EmbeddedDb(loadEngine().openMem());
  }

  #handle(): RawHandle {
    if (!this.#raw) throw new ClosedError("closed handle");
    return this.#raw;
  }

  /** Run one command; argv[0] is the verb. Returns the decoded Reply — a
   * protocol -ERR arrives as a Reply of kind "error" (data, not a throw). */
  cmd(...argv: Data[]): Reply {
    const bytes = this.#handle().cmdRaw(argv.map(toBytes));
    if (!bytes) throw new ClosedError("empty reply");
    return decodeReply(bytes);
  }

  /** Scalar fast GET, or null on miss/expired (§5.2). */
  getScalar(key: Data): Uint8Array | null {
    return this.#handle().getScalar(toBytes(key));
  }
  /** Scalar fast SET (ttlMs 0 = no TTL). */
  setScalar(key: Data, val: Data, ttlMs = 0): void {
    this.#handle().setScalar(toBytes(key), toBytes(val), ttlMs);
  }

  subscribe(channel: Data): EmbSub {
    return new EmbSub(this.#handle().subscribe(toBytes(channel), false));
  }
  psubscribe(pattern: Data): EmbSub {
    return new EmbSub(this.#handle().subscribe(toBytes(pattern), true));
  }

  close(): void {
    this.#raw?.close();
    this.#raw = null;
  }

  /** Subscribe to a raw channel/pattern (used by the Subscriber layer). */
  subBytes(name: Uint8Array, pattern: boolean): EmbSub {
    return new EmbSub(this.#handle().subscribe(name, pattern));
  }
}

/** The engine version string. */
export function version(): string {
  return loadEngine().version();
}
/** The runtime C ABI version. */
export function abi(): number {
  return loadEngine().abi();
}

// --- process-global embedded registry (contract §1.3) -------------------
//
// A URL-keyed refcounted map: two connect() calls with the same mem://<name>
// or file://<path> share one store (and one pub/sub bus). The entry evicts
// when the last handle drops.

interface Shared {
  db: EmbeddedDb;
  refs: number;
}
const registry = new Map<string, Shared>();

/** Open (or share) the embedded store for a target; returns [db, key]. */
export function resolveStore(t: Target): [EmbeddedDb, string] {
  const key = registryKey(t);
  if (key !== "") {
    const s = registry.get(key);
    if (s) {
      s.refs++;
      return [s.db, key];
    }
  }
  const db = openForTarget(t);
  if (key === "") return [db, ""];
  registry.set(key, { db, refs: 1 });
  return [db, key];
}

/** Drop one reference to a shared store, closing it when the last goes. */
export function releaseStore(key: string, db: EmbeddedDb): void {
  if (key === "") {
    db.close();
    return;
  }
  const s = registry.get(key);
  if (!s) return;
  s.refs--;
  if (s.refs <= 0) {
    s.db.close();
    registry.delete(key);
  }
}

function openForTarget(t: Target): EmbeddedDb {
  if (t.kind === "file") return EmbeddedDb.open(t.path);
  return EmbeddedDb.openMem();
}

File: bindings/cpp/src/embedded.cpp
#include "kevy/embedded.hpp"

#include <vector>

#include "kevy.h"
#include "kevy/errors.hpp"
#include "kevy/resp.hpp"

namespace kevy {
namespace {

// Decode a kevy-owned KevyBuf into a Reply, then free it (exactly once).
Reply take_reply(KevyBuf buf) {
  if (buf.ptr == nullptr || buf.len == 0) {
    kevy_buf_free(buf.ptr, buf.len, buf.cap);
    throw ProtocolError("kevy: empty reply from embedded cmd");
  }
  try {
    Reply r = resp::decode_reply(buf.ptr, buf.len);
    kevy_buf_free(buf.ptr, buf.len, buf.cap);
    return r;
  } catch (...) {
    kevy_buf_free(buf.ptr, buf.len, buf.cap);
    throw;
  }
}

const uint8_t* byte_ptr(std::string_view s) {
  // The C ABI wants a non-null pointer even for length 0.
  static const uint8_t kEmpty = 0;
  return s.empty() ? &kEmpty : reinterpret_cast<const uint8_t*>(s.data());
}

}  // namespace

// --- Subscription -------------------------------------------------------

Subscription& Subscription::operator=(Subscription&& o) noexcept {
  if (this != &o) {
    close();
    sub_ = o.sub_;
    o.sub_ = nullptr;
  }
  return *this;
}

std::optional<Reply> Subscription::next() {
  if (sub_ == nullptr) throw ClosedError();
  KevyBuf out{};
  int32_t rc = kevy_sub_next(sub_, &out);
  if (rc < 0) throw ClosedError();
  if (rc == 0) return std::nullopt;
  return take_reply(out);
}

std::optional<Reply> Subscription::wait(uint64_t timeout_ms) {
  if (sub_ == nullptr) throw ClosedError();
  KevyBuf out{};
  int32_t rc = kevy_sub_wait(sub_, timeout_ms, &out);
  if (rc < 0) throw ClosedError();
  if (rc == 0) return std::nullopt;
  return take_reply(out);
}

void Subscription::close() {
  if (sub_ != nullptr) {
    kevy_sub_close(sub_);
    sub_ = nullptr;
  }
}

// --- EmbeddedStore ------------------------------------------------------

EmbeddedStore EmbeddedStore::open(std::string_view dir) {
  KevyDb* db = kevy_open(byte_ptr(dir), dir.size());
  if (db == nullptr) throw IoError("kevy: open failed");
  return EmbeddedStore(db);
}

EmbeddedStore EmbeddedStore::open_mem() {
  KevyDb* db = kevy_open_mem();
  if (db == nullptr) throw IoError("kevy: open_mem failed");
  return EmbeddedStore(db);
}

EmbeddedStore& EmbeddedStore::operator=(EmbeddedStore&& o) noexcept {
  if (this != &o) {
    close();
    db_ = o.db_;
    o.db_ = nullptr;
  }
  return *this;
}

void EmbeddedStore::close() {
  if (db_ != nullptr) {
    kevy_close(db_);
    db_ = nullptr;
  }
}

Reply EmbeddedStore::cmd(const std::vector<std::string_view>& argv) {
  if (db_ == nullptr) throw ClosedError();
  if (argv.empty()) throw InvalidInputError("kevy: empty argv");
  std::vector<const uint8_t*> ptrs(argv.size());
  std::vector<size_t> lens(argv.size());
  for (size_t i = 0; i < argv.size(); i++) {
    ptrs[i] = byte_ptr(argv[i]);
    lens[i] = argv[i].size();
  }
  KevyBuf out{};
  int32_t rc = kevy_cmd(db_, argv.size(), ptrs.data(), lens.data(), &out);
  if (rc != 0) throw ProtocolError("kevy: kevy_cmd misuse");
  return take_reply(out);
}

// --- SharedValue --------------------------------------------------------

SharedValue& SharedValue::operator=(SharedValue&& o) noexcept {
  if (this != &o) {
    release();
    ptr_ = o.ptr_;
    len_ = o.len_;
    cap_ = o.cap_;
    owns_ = o.owns_;
    o.owns_ = false;
    o.ptr_ = nullptr;
  }
  return *this;
}

void SharedValue::release() {
  if (owns_) {
    kevy_buf_free_shared(ptr_, len_, cap_);
    owns_ = false;
  }
}

// --- get / get_view -----------------------------------------------------

std::optional<SharedValue> EmbeddedStore::get_view(std::string_view key) {
  if (db_ == nullptr) throw ClosedError();
  KevyBuf out{};
  // Zero-copy shared lane: a bulk value is an Arc clone (refcount bump, no
  // byte copy) that SharedValue owns and views; the shared free drops the Arc
  // in its destructor.
  int32_t rc = kevy_get_shared(db_, byte_ptr(key), key.size(), &out);
  if (rc < 0) throw ProtocolError("kevy: kevy_get_shared misuse");
  if (rc == 0) return std::nullopt;
  return SharedValue(out.ptr, out.len, out.cap);
}

std::optional<std::string> EmbeddedStore::get(std::string_view key) {
  // The copying convenience over the shared lane; view() is empty (not a
  // dangling range) for a 0-length value, so std::string(view) is UB-free.
  std::optional<SharedValue> sv = get_view(key);
  if (!sv.has_value()) return std::nullopt;
  return std::string(sv->view());
}

void EmbeddedStore::set(std::string_view key, std::string_view value, uint64_t ttl_ms) {
  if (db_ == nullptr) throw ClosedError();
  int32_t rc = kevy_set(db_, byte_ptr(key), key.size(), byte_ptr(value), value.size(), ttl_ms);
  if (rc < 0) throw IoError("kevy: kevy_set misuse or storage error");
}

Subscription EmbeddedStore::subscribe(std::string_view channel) {
  if (db_ == nullptr) throw ClosedError();
  KevySub* s = kevy_subscribe(db_, byte_ptr(channel), channel.size());
  if (s == nullptr) throw IoError("kevy: subscribe failed");
  return Subscription(s);
}

Subscription EmbeddedStore::psubscribe(std::string_view pattern) {
  if (db_ == nullptr) throw ClosedError();
  KevySub* s = kevy_psubscribe(db_, byte_ptr(pattern), pattern.size());
  if (s == nullptr) throw IoError("kevy: psubscribe failed");
  return Subscription(s);
}

std::string EmbeddedStore::version() { return std::string(kevy_version()); }
uint32_t EmbeddedStore::abi() { return kevy_abi(); }

}  // namespace kevy
